### ISSUE TITLE
Trailing punctuation causes multiple results when using Browse Alphabetically with authors/topics 

### ISSUE THREAD

**Issue by gibsonjc** [CARLI/vufind#287]
Author:
![screen shot 2018-01-12 at 3 44 01 pm](https://user-images.githubusercontent.com/17598357/34896598-0017b812-f7b0-11e7-91e9-78abee0ec47a.png)

Topic:
![screen shot 2018-01-12 at 3 45 19 pm](https://user-images.githubusercontent.com/17598357/34896637-1f95929a-f7b0-11e7-9559-cfedc63c9be6.png)

**Comment by gibsonjc** [comment]
LFC - open call 12/15/2017

**Comment by gibsonjc** [comment]
@cedelis We talked about this in email a little bit on August 17, 2017 (Subj: Library data with periods at the end).  At the time, one idea was to look into stripping punctuation from the end of the strings when Alphabrowse makes its index so it didn't see them as separate entries.  

https://vufind.org/wiki/indexing:alphabetical_heading_browse 
has a page on Alphabrowse that talks about normalization, and I think you've done some work for call numbers already.  

I don't see the same problem that we have in the VuFind Demo or other VuFind catalogs like UofC, so I assume there is a known solution of some kind: 
https://vufind.org/demo/Alphabrowse/Home?source=topic&from=automobiles
https://catalog.lib.uchicago.edu/vufind/Alphabrowse/Home?from=automobiles&source=topic

**Comment by cedelis** [comment]
@gibsonjc Yes, there appears to be a known solution. SolrMarc has added new features to its scripting language:

https://github.com/solrmarc/solrmarc/wiki/Index-Specification-Language

<pre>
join    or  join(" : ") -- concatenate the matching subfields producing one output line per matching datafield
separate                -- do not concatenate matching subfields, instead output one line per matching subfield
format                  --
substring(start, end)   -- extract only a range of the data in the matching subfield
cleanEach               -- strip punctuation from each matching subfield
cleanEnd                -- strip punctuation after the subfields are concatenated
clean                   -- equal to specifying both cleanEnd and cleanEach
stripAccent             -- strip accent marks and diacritics from characters
stripPunct              -- strip all punctuation everywhere
stripInd2               -- strip a number of characters from the $a subfield equal to the digit in indicator2
toUpper                 -- convert all letters to uppercase
toLower                 -- convert all letters to lowercase
titleSortUpper          -- equal to cleanEach, cleanEnd, stripAccent, stripPunct, stripInd2, toUpper
titleSortLower          -- equal to cleanEach, cleanEnd, stripAccent, stripPunct, stripInd2, toLower
untrimmed               -- cause it to NOT call String.trim()
</pre>

I believe we can apply the "clean" custom indexing method to the author and topic fields.

FYI, the author fields used in browse currently consist of:

<pre>
author                = custom, getAuthorsFilteredByRelator(100abcd:700abcd,100,firstAuthorRoles)
author2               = custom, getAuthorsFilteredByRelator(700abcd:800abcd,700,secondAuthorRoles)
author_corporate      = custom, getAuthorsFilteredByRelator(110ab:111abc:410ab:411abc:710ab:711ab:810ab:811abc,110:111:710:711,firstAuthorRoles|secondAuthorRoles)
</pre>

...and the topic field used in browse consists of:

<pre>
topic = custom, getAllSubfields(600:610:611:630:650:653:656:690:696, " ")
</pre>


I would need to add "clean" to each spec above, e.g.,

<pre>
topic = custom, getAllSubfields(600:610:611:630:650:653:656:690:696, " "), clean
</pre>

Unfortunately, this means we'll need to schedule a complete reindex in our local catalogs. In the meantime, I will begin indexing in development. If we like what we see, we can then schedule a reindex in the future. 

Edit (Tue Apr 17 14:46): Hmm... it seems that the new features (e.g., clean) are not implemented in conjunction with custom handlers. Therefore, I think I need to simply truncate the characters myself in an overriden version of the original getAuthorsFilteredByRelator() methods. That's too bad.

**Comment by cedelis** [comment]
Turns out we needed to amend our indexing scripts to trim trailing punctuation on the author_browse and topic_browse Solr fields. We also needed to adjust the authority Solr fields, heading, use_for, and see_also.

Since this requires a complete re-indexing of the local catalog, for now, this is available only on our development server (which is firewalled). Eventually, this will be done on TEST, and then finally on PRODUCTION.
